language-plutus-core/src/Language/PlutusCore/Universe/Core.hs

```diff
@@ -0,0 +1,348 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DefaultSignatures     #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE TemplateHaskell       #-}
+
+module Language.PlutusCore.Universe.Core
```

Comment:
It's interesting that basically all of this module is generic, and not related to PLC at all. It reminds me a lot of the whole universe/coding trick that people use in dependently typed languages (not a coincidence, I'm sure). Maybe there's a library in here.

Reply:
> It's interesting that basically all of this module is generic, and not related to PLC at all.

Yeah, that's intentional.

> Maybe there's a library in here.

There is [`Data.Dependent.Sum`](https://hackage.haskell.org/package/dependent-sum-0.6.2.0/docs/Data-Dependent-Sum.html), but I didn't want to depend on any library just not to restrict myself. I think we have a nicer story for deriving. Probably once we settle on everything (pun intended), it'll make sense to create our own library.

Reply:
> It's interesting that basically all of this module is generic, and not related to PLC at all.

Yeah, that's intentional.

> Maybe there's a library in here.

There is [`Data.Dependent.Sum`](https://hackage.haskell.org/package/dependent-sum-0.6.2.0/docs/Data-Dependent-Sum.html), but I didn't want to depend on any library just not to restrict myself. I think we have a nicer story for deriving. Probably once we settle on everything (pun intended), it'll make sense to create our own library.

Reply:
I mean, we do depend on `dependent-sum` to get `GEq`, `GShow`, etc, but we don't depend on the actual `DSum` data type.

Reply:
I mean, we do depend on `dependent-sum` to get `GEq`, `GShow`, etc, but we don't depend on the actual `DSum` data type.